cmd/opampsupervisor/e2e_test.go

```diff
@@ -0,0 +1,389 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+//go:build e2e
+// +build e2e
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"io"
+	"log"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path"
+	"runtime"
+	"strings"
+	"sync/atomic"
+	"testing"
+	"text/template"
+	"time"
+
+	"github.com/open-telemetry/opamp-go/protobufs"
+	"github.com/open-telemetry/opamp-go/server"
+	"github.com/open-telemetry/opamp-go/server/types"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/cmd/opampsupervisor/supervisor"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+)
+
+type testLogger struct {
+	t *testing.T
+}
+
+func (tl testLogger) Debugf(format string, args ...any) {
+	tl.t.Logf(format, args...)
+}
+
+func (tl testLogger) Errorf(format string, args ...any) {
+	tl.t.Logf(format, args...)
+}
+
+func defaultConnectingHandler(connecti
+	onCallbacks server.ConnectionCallbacksStruct) func(request *http.Request) types.ConnectionResponse {
+	return func(request *http.Request) types.ConnectionResponse {
+		return types.ConnectionResponse{
+			Accept:              true,
+			ConnectionCallbacks: connectionCallbacks,
+		}
+	}
+}
+
+// onConnectingFuncFactory is a function that will be given to server.CallbacksStruct as
+// OnConnectingFunc. This allows changing the ConnectionCallbacks both from the newOpAMPServer
+// caller and inside of newOpAMP Server, and for custom implementations of the value for `Accept`
+// in types.ConnectionResponse.
+type onConnectingFuncFactory func(connectionCallbacks server.ConnectionCallbacksStruct) func(request *http.Request) types.ConnectionResponse
+
+type testingOpAMPServer struct {
+	addr                string
+	supervisorConnected chan bool
+	sendToSupervisor    func(*protobufs.ServerToAgent)
+}
+
+func newOpAMPServer(t *testing.T, connectingCallback onConnectingFuncFactory, callbacks server.ConnectionCallbacksStruct) *testingOpAMPServer {
```

Comment:
From what I see we only call `newOpAMPServer` with `defaultConnectingHandler`. Do we need `connectingCallback` to be a parameter?

Reply:
I didn't add tests for it as part of this PR, but I was thinking this could be used to test the Supervisor's behavior if the server rejects its connection attempt. Maybe there's a better way to cover this single case, but leaving the behavior configurable with a default function seemed like the most straightforward solution for now.